src/backend/distributed/executor/citus_custom_scan.c

```diff
@@ -801,3 +817,53 @@ IsCitusCustomScan(Plan *plan)
 
 	return true;
 }
+
+
+/*
+ * In a Job, given a list of relations, if all them belong to the same
+ * colocation group, the Job's colocation ID is set to the group ID, else,
+ * it will be set to INVALID_COLOCATION_ID.
+ */
+static void
+SetJobColocationId(Job *job)
+{
+	uint32 jobColocationId = INVALID_COLOCATION_ID;
+
+	if (job->partitionKeyValue == NULL)
+	{
+		/* If the shard key is not known, colocationId is irrelevant */
+		job->colocationId = INVALID_COLOCATION_ID;
+		return;
+	}
+
+	List *rangeTableList = job->jobQuery->rtable;
+	ListCell *rangeTableCell = NULL;
+	foreach(rangeTableCell, rangeTableList)
+	{
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
+		Oid relationId = rangeTableEntry->relid;
+
+		if (!IsCitusTable(relationId))
+		{
+			/* ignore the non distributed table */
+			continue;
+		}
+
+		uint32 colocationId = TableColocationId(relationId);
+		Assert(colocationId != INVALID_COLOCATION_ID);
```

Comment:
I hit this assert failure when running the regression tests on my laptop

```sql
CREATE TABLE source_table(f1 int, key composite_key_type, value int, mapped_key composite_key_type);
 SELECT create_distributed_table('source_table', 'key', 'range');
  create_distributed_table
 --------------------------

 (1 row)

 CALL public.create_range_partitioned_shards('source_table', '{"(0,a)","(25,a)"}','{"(24,z)","(49,z)"}');
 INSERT INTO source_table VALUES (0, (0, 'a'), 1, (0, 'a'));
```
From the log:
```
TRAP: FailedAssertion("colocationId != INVALID_COLOCATION_ID", File: "executor/citus_custom_scan.c", Line: 853, PID: 21173)
postgres: postgres regression 127.0.0.1(51416) INSERT(ExceptionalCondition+0xa5)[0x562463da7851]
/home/marco/pgsql/14/lib/citus.so(+0x9dff3)[0x7fa8a5047ff3]
/home/marco/pgsql/14/lib/citus.so(+0x9e2fb)[0x7fa8a50482fb]
postgres: postgres regression 127.0.0.1(51416) INSERT(ExecInitCustomScan+0xe3)[0x562463ab1084]
postgres: postgres regression 127.0.0.1(51416) INSERT(ExecInitNode+0x2f4)[0x562463a9b32f]
postgres: postgres regression 127.0.0.1(51416) INSERT(+0x2f21dd)[0x562463a941dd]
postgres: postgres regression 127.0.0.1(51416) INSERT(standard_ExecutorStart+0x132)[0x562463a943c3]
/home/marco/pgsql/14/lib/citus.so(CitusExecutorStart+0x59)[0x7fa8a504e093]
/home/marco/pgsql/14/lib/pg_stat_statements.so(+0x4e9b)[0x7fa8a4fa2e9b]
```

Reply:
Let me repro this.

Reply:
I need to rethink this assertion
**Range distribution**
postgres=# select logicalrelid, colocationid from pg_dist_partition;
 logicalrelid | colocationid
--------------+--------------
 source_table |            **0**
(1 row)

postgres=# select * from pg_dist_shard;
 logicalrelid | shardid | shardstorage | shardminvalue | shardmaxvalue
--------------+---------+--------------+---------------+---------------
 source_table |  102008 | t            | (0,a)         | (24,z)
 source_table |  102009 | t            | (25,a)        | (49,z)
(2 rows)

**Hash distribution**
postgres=# select logicalrelid, colocationid from pg_dist_partition;
 logicalrelid | colocationid
--------------+--------------
 source_table |           **35**

postgres=# select * from pg_dist_shard where shardid > 102073;
 logicalrelid | shardid | shardstorage | shardminvalue | shardmaxvalue
--------------+---------+--------------+---------------+---------------
 source_table |  102074 | t            | -2013265920   | -1879048193
 source_table |  102075 | t            | -1879048192   | -1744830465
 source_table |  102076 | t            | -1744830464   | -1610612737
 source_table |  102077 | t            | -1610612736   | -1476395009
 source_table |  102078 | t            | -1476395008   | -1342177281

Reply:
As colocate_with option is only supported for hash distributed tables and reference tables, the assert is invalid, removed it.